### ISSUE TITLE
Support for 'tags' in postings

### ISSUE THREAD

**Issue by skalk** [genodelabs/genodians.org#11]
John J. Karcher brought up this feature request in a posting to the Genode mailing list on 31th March 2019 with the subject "Genodians.org Feature Requests". He said:
> Filter Tags - I think it would be helpful to add "tags" to the
> posts to make it easier to find items on a certain subject.  I would recommend
> defining the format and adding the tag data as soon as possible, even if the
> feature isn't implemented for a while - it's always harder to go back over a
> lot of data than to enter it as the records are added.
@nfeske proposed to use gosh's annotations to specify tags, like
! | storage performance cryptography
with the restriction to have one-word tags only.
@m-stein further proposed to use an annotation like this:
! | tags: "storage" "performance" "symmetric cryptography"
to distinguish tags from normal annotations and to have more than one-word tags.

**Comment by m-stein** [comment]
@skalk Thanks for moving this to an issue!

**Comment by jjkarcher** [comment]
FWIW, I'm inclined toward using the "tags:" keyword, a la @m-stein.  The quoted string idea would also be more flexible (although you can get a similar result with underscores, it isn't as friendly for end-users).

**Comment by nfeske** [comment]
Let us consider that the tag names will be used as sub directory or file names within the html directory. Hence, they should not contain whitespace (BTW, twitter hash tags also don't have any whitespace). In my opinion, tags should be concise anyway and we should not go overboard with them. So the limitation to a single word reinforces the conciseness, which is good.

I question the usefulness of adding bureaucracy with the "tags:" annotation. What concrete benefit is gained by that? It would only make sense if we envision other forms of extensibility. But we are talking about a blog here, not a Wikipedia replacement. On the downside, @m-stein's suggestion introduces uncertainties. What should the tool do when the author misspells "tags:" as "tags;"? Adding syntax (without an actual need) is a bad idea in my opinion.

**Comment by skalk** [comment]
I see the issue with whitespaces, and I think if there are word combinations absolutely necessary to have, you can use an underscore as emergency solution like @jjkarcher mentioned.
I agree that we should not complicate stuff here, because it is a micro-blogging solution not Wordpress, although I do not get the "downside" argument with the uncertain syntax here. In case of misspelling the tool would handle tags as normal annotation instead of linking tags. The argument of misspelling does also apply if you accidentally type `\` instead of `|` as first character of the annotation.
But I agree that annotations are somethings rarely used in a micro-blogging solution anyway, so why should we support them at genodians.org except for tags. 

As @m-stein suggested before, let's finally reach a consensus about the convention in the gosh files in the first place, so that every author can already apply to it, before we incorporate a solution within the HTML frontend.

So how about placing an annotation at the end of the file, like initially suggested. To not confuse readers of the pure text posts, authors can also add a gosh comment before that line:
```
; tags used by this post
| storage performance symmetric_cryptography
```

**Comment by nfeske** [comment]
@skalk I agree with everything you just said.

The suggestion at the end of your post is beautiful because it addresses @m-stein's wish to be more explicit while at the same time giving me the freedom to be concise (leaving out the line with the semicolon).

**Comment by m-stein** [comment]
I totally agree with @skalk. Although I'd prefer a mandatory annotation of what the tag line means and don't see it as overly bureaucratic, I'm fine with the solution @skalk suggested.

**Comment by jschlatow** [comment]
Working at #34 and having the Makefile already in my mental cache, I gave this issue a shot as well. Commit a0c7d6d implements the idea discussed above. Note that this further increases the processing time so this should probably be merged only after improving the performance issue.

**Comment by chelmuth** [comment]
I tested your topic branch but `make` fails like follows (with and without topics added to my latest article).

```
generate html/chelmuth/author ...
generate html/chelmuth/2022-11-24-u7411-up-and-running ...
generate html/chelmuth/2020-03-16-lets-encrypt ...
generate html/chelmuth/2019-10-21-sculpt-elitebook ...
generate html/chelmuth/2019-06-07-terminal-palette ...
generate html/chelmuth/2019-03-13-powerplug ...
generate html/chelmuth/2019-01-31-network-dev ...
generate html/chelmuth/2019-01-16-test-machine ...
generate html/summary/chelmuth/2022-11-24-u7411-up-and-running ...
generate html/summary/chelmuth/2020-03-16-lets-encrypt ...
generate html/summary/chelmuth/2019-10-21-sculpt-elitebook ...
generate html/summary/chelmuth/2019-06-07-terminal-palette ...
generate html/summary/chelmuth/2019-03-13-powerplug ...
generate html/summary/chelmuth/2019-01-31-network-dev ...
generate html/summary/chelmuth/2019-01-16-test-machine ...
generate html/index ...
cat: html/topics: No such file or directory
Makefile:134: recipe for target 'html/index' failed
make: *** [html/index] Error 1
```

Note, I put just *genodian-chelmuth.git* into *content/*.

**Comment by chelmuth** [comment]
The following patch moves *Topics* below external links (which seems better with a growing list of topics) and fixes the dependency of front page on the topics snippet.

```diff
+++ b/Makefile
@@ -121,6 +121,9 @@ html/topics: $(foreach T,$(TOPICS),html/topics-$T)
 # make topic pages depend on article summaries
 $(foreach T,$(TOPICS),$(eval html/topics-$T: $(foreach P,${TOPIC_POSTINGS($T)},$(addprefix html/summary/,$P))))
 
+# front page depends on topics snippet
+html/index: html/topics
+
 # archive page depends on summary snippets
 html/archive: $(addprefix html/summary/,$(REV_POSTINGS))
 
@@ -170,10 +173,10 @@ html/index html/archive $(addprefix html/topics-,$(TOPICS)):
 	     "          </div> <!-- menu-inner -->\n" \
 	     "        </div> <!-- authors menu -->\n" \
 	     "      </div> <!-- authors-small -->" >> $@
+	cat style/external-links-menu >> $@
 	$(if $(filter-out html/topics-%,$@), \
 	  cat html/topics >> $@;)
-	cat style/external-links-menu \
-	    style/footer >> $@
+	cat style/footer >> $@
 
 html/rss:
 	$(MSG)
```

**Comment by jschlatow** [comment]
Thanks for testing. Your patch looks good to me.

**Comment by chelmuth** [comment]
So, what's your stance? May we merge the series despite the performance impact?

**Comment by jschlatow** [comment]
I'm not sure how strongly the Genodians VM will be affected. Under what VMM is it running?

**Comment by chelmuth** [comment]
> I'm not sure how strongly the Genodians VM will be affected. Under what VMM is it running?

It's running genode/nova in Linux/KVM.

**Comment by jschlatow** [comment]
I re-tested run/genodians.run with KERNEL=nova. Enabling KVM does the trick, i.e. when I add `-enable-kvm` to the qemu arguments, the generation is MUCH faster. When I only disable all authors but myself, the generation took 2min 20s without KVM. With KVM it took merely 10s.

When I enable all authors, everything is generated in less than 2min. I think this is good enough.

Note that I had to modify some RAM quota (see c8356c6).
